### DISCUSSION TITLE
What does atomicly really mean

### DISCUSSION THREAD

**Discussion by hput** [psy0rz/zfs_autobackup#130] [Q&A] [answered]
I cannot find a definition for the word `atomicly' used in our Readme.  Maybe its supposed to be `atomically', either way what the heck does it mean.

In the README under FEATURES we have:  "Creates consistent snapshots. (takes all snapshots at once, atomicly.)"

Just on the surface it sounds a bit ridiculous "takes all snapshots at once"; alone sounds miraculous.  How can you do that with no process or string of events.  Maybe its done "atomicly" But again what in the world can that possibly mean.  Unless its another word for `magically'

Sorry for sounding a bit negative but that just rattles my brain a bit.

**Comment by psy0rz** [comment, marked as answer]
edit: Look at @Scrin 's comment :)

its a technical term. look in the zfs manual:

> Snapshots are taken atomically, so that all recursive snapshots correspond to the same moment in time.

**Reply by hput** [reply]
I'm sure I'm sounding like some kind of hard headed knuckle head, but seeing its use in zfs manual in 2 places, both the one you quoted and one more; hasn't made me understand how things can just happen all at once.
Surely it takes some time to run something like:  zfs snap -r p0@$(date +"%y%m%d%H%M%S").

In this case p0 pool has only a few zfs fs on it (my count is 8) so it happens quickly but not "all at once"

     time zfs snap -r p0@$(date +"%y%m%d%H%M%S")

real    0m0.192s
user    0m0.002s
sys     0m0.014s

So, fast yes, but not "atomically"
Were that pool populated with 100+ zfs fs, I'm pretty sure the time factor would be a bit larger.

**Reply by Scrin** [reply]
The snapshots are taken by the zfs kernel module (which is responsible for actually "doing" most of the "zfs operations", such as the ones triggered by the zfs or zpool commands/programs). Essentially when the kernel module gets a call to make an atomic, recursive snapshot, it will synchronize the actual snapshotting such that all writes are frozen until the snapshotting has completed, which means the state of all the datasets being snapshot will correspond to the same point in time, since no writes are happening during that time so none of the datasets change.

You are right that having a significant number of datasets (like thousands) while heavy write operations are happening, you can see a dent in the write throughput during snapshotting due to all writes being frozen during that time, and this time will depend on the amount of datasets (among other things).

Do note that the durations observed by the `time` utility do not correspond to the duration of the "freeze"; the kernel module does some "preparative work" before the "freeze" and snapshotting, such as creating the list of datasets to snapshot, and the duration it takes to do these is not a constant.

**Reply by hput** [reply]
Thank you sir for the well organized  and thoroughly helpful answer!
I was beginning to think zfs users were some kind of religious cult [JOKE]

Seriously, that answer really pushed me into a higher and much clearer understanding.  I hadn't even considered how the kernel might be helping thing along
